python/pyarrow/types.pxi

```diff
@@ -114,7 +123,14 @@ cdef void* _as_c_pointer(v, allow_null=False) except *:
             "Arrow library", UserWarning, stacklevel=2)
         c_ptr = <void*> <uintptr_t > v
     elif PyCapsule_CheckExact(v):
-        c_ptr = PyCapsule_GetPointer(v, NULL)
+        capsule_name = PyCapsule_GetName(v)
+        if capsule_name == NULL or capsule_name == "r_extptr":
```

Comment:
Did you check that `capsule_name == "r_extptr"` works as expected?

Reply:
```suggestion
        if capsule_name == NULL or capsule_name == b"r_extptr":
```

I suspect it should be compared with bytes when comparing directly (our `frombytes` helper converts from the bytes to string)